server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionListenerTests.java

```diff
@@ -59,4 +67,84 @@ public void testShouldDelegateFailureImmediately() {
 
         assertThat(completed.get(), equalTo(true));
     }
+
+    public void testShouldExecuteWithCorrectContext() {
+
+        var context = new ThreadContext(Settings.EMPTY);
+
+        // should not be changed after the listener is created
+        context.putHeader("header", "root");// ensure this is visible in the listener
+        context.addResponseHeader("header", "1");
+
+        var listener = new AllocationActionListener<>(ActionListener.wrap(ignore -> {
+            assertThat(context.getResponseHeaders().get("header"), containsInAnyOrder("1", "3", "4"));
+            assertThat(context.getHeader("header"), equalTo("root"));
+        }, exception -> { throw new AssertionError("Should not fail in test"); }), context);
+
+        // this header should be ignored as it is added after context is captured
+        context.addResponseHeader("header", "2");
+
+        // reroute is executed for multiple changes so its headers should be ignored
```

Comment:
nit: duplicated comment (should this one be about the CS update listener instead?)

```suggestion
        // headers for clusterStateUpdate listener are captured
```